src/pools/D3MAaveDaiPool.sol

```diff
@@ -0,0 +1,161 @@
+        require(adai_ != address(0), "D3MAaveDaiPool/invalid-adai");
+        require(stableDebt_ != address(0), "D3MAaveDaiPool/invalid-stableDebt");
+        require(variableDebt_ != address(0), "D3MAaveDaiPool/invalid-variableDebt");
+        require(interestStrategy_ != address(0), "D3MAaveDaiPool/invalid-interestStrategy");
+
+        adai = adai_;
+        stableDebt = ShareTokenLike(stableDebt_);
+        variableDebt = ShareTokenLike(variableDebt_);
+        interestStrategy = interestStrategy_;
+        rewardsClaimer = RewardsClaimerLike(_rewardsClaimer);
+
+        ShareTokenLike(adai_).approve(pool_, type(uint256).max);
+        TokenLike(dai_).approve(pool_, type(uint256).max);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    // --- Math ---
+    function _add(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x + y) >= x, "DssDirectDepositHub/overflow");
+    }
+    function _mul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require(y == 0 || (z = x * y) / y == x, "D3MAaveDaiPool/overflow");
+    }
+    function _rdiv(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _mul(x, RAY) / y;
+    }
+    function _min(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x <= y ? x : y;
+    }
+
+    // --- Admin ---
+    function file(bytes32 what, address data) external auth {
+        require(live == 1, "D3MTestPool/no-file-not-live");
+
+        if (what == "king") king = data;
+        else revert("D3MPoolBase/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    function validTarget() external view override returns (bool) {
+        (,,,,,,,,,, address strategy,) = LendingPoolLike(pool).getReserveData(address(asset));
+        return strategy == interestStrategy;
+    }
+
+    // Deposits Dai to Aave in exchange for adai which gets sent to the msg.sender
+    // Aave: https://docs.aave.com/developers/v/2.0/the-core-protocol/lendingpool#deposit
+    function deposit(uint256 amt) external override auth {
+        uint256 scaledPrev = ShareTokenLike(adai).scaledBalanceOf(address(this));
+
+        LendingPoolLike(pool).deposit(address(asset), amt, address(this), 0);
+
+        // Verify the correct amount of adai shows up
+        uint256 interestIndex = LendingPoolLike(pool).getReserveNormalizedIncome(address(asset));
+        uint256 scaledAmount = _rdiv(amt, interestIndex);
+        require(ShareTokenLike(adai).scaledBalanceOf(address(this)) >= _add(scaledPrev, scaledAmount), "D3MAaveDaiPool/incorrect-share-credit");
+    }
+
+    // Withdraws Dai from Aave in exchange for adai
+    // Aave: https://docs.aave.com/developers/v/2.0/the-core-protocol/lendingpool#withdraw
+    function withdraw(uint256 amt) external override auth {
+        LendingPoolLike(pool).withdraw(address(asset), amt, address(hub));
```

Comment:
I would switch `address(hub)` to `msg.sender`. We can drop the `hub` reference after that.

Reply:
Well the `hub` is still necessary for the hoping part in the constructor, but it is true that could be removed as immutable.

Reply:
I removed it as immutable and left the constructor `hope`.  I could create a function for that so we don't have to pass it at all, but I would prefer to leave it like this especially if we do away with the base contracts in favor of interfaces